### PR TITLE
Add progress info for the Osquery Flow (Backend)

### PR DESCRIPTION
**Pull request by simstoykov**
Report the first 10 rows, together with the total number of rows as "Progress".
This PR only contains the backend changes.

### CODE REVIEW

File: grr/server/grr_response_server/flows/general/osquery.py [hunk 1]
```diff
@@ -19,17 +23,43 @@ class OsqueryFlow(flow_base.FlowBase):
...
         server_stubs.Osquery,
         request=self.args,
         next_state=compatibility.GetName(self.Process))
 
-  def Process(self, responses):
+  def Process(self, responses: flow_responses.Responses[rdf_osquery.OsqueryResult]) -> None:
     if not responses.success:
       raise flow_base.FlowError(responses.status)
 
+    self.UpdateProgress(responses)
```
Comment (mbushkov) [inline grr/server/grr_response_server/flows/general/osquery.py:41]: I would make UpdateProgress protected: _UpdateProgress.

File: grr/server/grr_response_server/flows/general/osquery.py [hunk 2]
```diff
@@ -19,17 +23,43 @@ class OsqueryFlow(flow_base.FlowBase):
...
+
     for response in responses:
       self.SendReply(response)
+
+  def GetProgress(self) -> rdf_osquery.OsqueryProgress:
+    return self.state.progress
+
+  def UpdateProgress(self, responses: flow_responses.Responses[rdf_osquery.OsqueryResult]) -> None:
```
Comment (mbushkov) [inline grr/server/grr_response_server/flows/general/osquery.py:49]: Throughout the codebase autoformat with 80 characters max line length. Please use this line length limit throughout the file (do you think you can configure your autoformatter with the same setting?)

File: grr/server/grr_response_server/flows/general/osquery.py [hunk 3]
```diff
@@ -19,17 +23,43 @@ class OsqueryFlow(flow_base.FlowBase):
   behaviours = flow_base.BEHAVIOUR_BASIC
 
   args_type = rdf_osquery.OsqueryArgs
+  progress_type = rdf_osquery.OsqueryProgress
```
Comment (mbushkov) [inline grr/server/grr_response_server/flows/general/osquery.py:26]: I don't see a change in rdfvalues/osquery.py included in to this PR.

**Review by mbushkov — CHANGES_REQUESTED** [review]
Looks really good! Please see a few minor comments below.

File: grr/server/grr_response_server/flows/general/osquery_test.py [hunk 4]
```diff
@@ -148,47 +153,101 @@ def testFailure(self):
...
+        client_id=self.client_id,
+        token=self.token,
+        query=query)
+    return session_id
+
+  def _RunQuery(self, query):
+    session_id = self._NewTestFlowStartAndRun(query)
+    return flow_test_lib.GetFlowResults(self.client_id, session_id)
+
+  def _AssertTablesMatch(self, table: rdf_osquery.OsqueryTable, expected: List[OrderedDict]) -> None:
```
[the remark this answers is not in the document]
Comment (simstoykov) [reply grr/server/grr_response_server/flows/general/osquery_test.py:226]: I agree in this particular case - my code should be consistent with the rest of the code base.

In general, though, what are the benefits of that? I guess this might be a left-over from languages like C in which you need to declare things before using them.

If instead protected methods are put at the end, they will make way for the public methods to be first. The public methods are the interface exposed, so if anyone needs to look at the class they will be presented with the core high-level functionality first. This in my opinion is better than first exposing low-level helper functions?

**Comment by simstoykov** [comment]
@mbushkov can you PTAL

File: grr/server/grr_response_server/flows/general/osquery_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mbushkov) [reply grr/server/grr_response_server/flows/general/osquery_test.py]: > If instead protected methods are put at the end, they will make way for the public methods to be first. The public methods are the interface exposed, so if anyone needs to look at the class they will be presented with the core high-level functionality first. This in my opinion is better than first exposing low-level helper functions?

This is a good question. I don't have a strong opinion on this as I think that practical differences between different ways of grouping methods together are negligible. It's normally useful to have them grouped (even though sometimes it actually makes sense to put a protected method right next to the public method using it), but whether it's better to have public first, protected second or protected first, public second - feels like a subjective preference. Developers have different backgrounds, use different IDEs, etc.

I personally find it useful to have the implementation details grouped together, and since often private object properties are defined at the top of the class, I see the point of grouping them with private/protected methods. Then, when I read the class definition I skim through all the private/protected stuff, getting an idea of what is there, and then I can read through public implementation with that additional context in mind. Again, this is very subjective, so the only real guideline in this question is code consistency - whichever approach we choose, makes sense to stick to it.

**Review by mbushkov — APPROVED** [review]

File: grr/server/grr_response_server/flows/general/osquery_test.py [hunk 5]
```diff
@@ -147,48 +152,108 @@ def testFailure(self):
...
-    stdout = """
-    [
-      { "foo": "quux", "bar": "norf", "baz": "thud" },
-      { "foo": "blargh", "bar": "plugh", "baz": "ztesch" }
+    test_table = [
+      OrderedDict([ ("foo", "quux"), ("bar", "norf"), ("baz", "thud") ]),
+      OrderedDict([ ("foo", "blargh"), ("bar", "plugh"), ("baz", "ztesch") ]),
     ]
-    """
+    stdout = json.Dump(test_table)
```
Comment (panhania) [inline grr/server/grr_response_server/flows/general/osquery_test.py:209]: Why? This contains more logic and is more complicated (it has more structure, yes, but this is just test data, do we care?).

File: grr/server/grr_response_server/flows/general/osquery_test.py [hunk 6]
```diff
@@ -147,48 +152,108 @@ def testFailure(self):
...
-class FakeOsqueryFlowTest(flow_test_lib.FlowTestsBaseclass):
+def _CreateDictTable(column_number, row_number) -> List[OrderedDict]:
+  def CellValue(col_number, row_number):
+    return (f"col-{col_number}", f"col-{col_number}, row-{row_number}")
+
+  return [
+    OrderedDict(
+      CellValue(col_number, row_number) for col_number in range(column_number)
+    ) for row_number in range(row_number)
+  ]
```
Comment (panhania) [inline grr/server/grr_response_server/flows/general/osquery_test.py:163]: This makes the test code hard to follow—avoid using any helpers in the test code if possible. Here, you don't care about the values at all (whether they are unique or not) and the number of columns is not too high, so you can prepare a readable test data yourself, e.g. `[{ "foo": "quux", "baz": "thud", "bar": "norf" }] * 42`.

File: grr/server/grr_response_server/flows/general/osquery.py [hunk 7]
```diff
@@ -19,17 +32,41 @@ class OsqueryFlow(flow_base.FlowBase):
...
+
+  def _UpdateProgress(
+    self,
+    responses: flow_responses.Responses[rdf_osquery.OsqueryResult]
+  ) -> None:
+    # XXX: Less than TRUNCATED_ROWS_COUNT will be used if the original table is
```
Comment (panhania) [inline grr/server/grr_response_server/flows/general/osquery.py:41]: What is `XXX`?

File: grr/server/grr_response_server/flows/general/osquery.py [hunk 8]
```diff
@@ -19,17 +32,41 @@ class OsqueryFlow(flow_base.FlowBase):
...
+
     self.CallClient(
         server_stubs.Osquery,
         request=self.args,
         next_state=compatibility.GetName(self.Process))
 
-  def Process(self, responses):
+  def Process(
+      self,
+      responses: flow_responses.Responses[rdf_osquery.OsqueryResult]
```
Comment (panhania) [inline grr/server/grr_response_server/flows/general/osquery.py:61]: Missing trailing comma (it is needed to force auto-formatter not to fold the result type to the same line).

File: grr/server/grr_response_server/flows/general/osquery.py [hunk 9]
```diff
@@ -19,17 +32,41 @@ class OsqueryFlow(flow_base.FlowBase):
...
     if not responses.success:
       raise flow_base.FlowError(responses.status)
 
+    self._UpdateProgress(responses)
+
     for response in responses:
       self.SendReply(response)
+
+  def GetProgress(self) -> rdf_osquery.OsqueryProgress:
+    return self.state.progress
```
Comment (panhania) [inline grr/server/grr_response_server/flows/general/osquery.py:72]: This can be an accessor.

File: grr/server/grr_response_server/flows/general/osquery.py [hunk 10]
```diff
@@ -9,6 +9,19 @@
 from grr_response_core.lib.util import compatibility
 from grr_response_server import flow_base
 from grr_response_server import server_stubs
+from grr_response_server import flow_responses
+
+
+TRUNCATED_ROWS_COUNT = 10
+
+def _TruncateRows(table: rdf_osquery.OsqueryTable) -> rdf_osquery.OsqueryTable:
```
Comment (panhania) [inline grr/server/grr_response_server/flows/general/osquery.py:17]: I would move this to a method on `OsqueryTable`. If you do, then `Truncate` sounds very imperative, as if it was doing this update in-place. For example, in Python `foo.sort()` will sort in-place, but `sorted(foo)` will construct a new, sorted object. If you apply both this suggestions, then the code would look like:

```python
foo = bar.Truncated()
```
rather than

```python
bar = _TruncateTable(foo)
```

which is just prettier IMO.

I would also make the number of rows to truncate an optional parameter (otherwise it would be impossible to test it).

**Review by panhania — CHANGES_REQUESTED** [review]

File: grr/server/grr_response_server/flows/general/osquery_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (panhania) [reply grr/server/grr_response_server/flows/general/osquery_test.py]: > Then, when I read the class definition I skim through all the private/protected stuff, getting an idea of what is there, and then I can read through public implementation with that additional context in mind.

Here, I agree with @simstoykov. Why should you need to read the private and protected parts of the class? Public interface is all you in general should care about. Sometimes it is not (if the code is poorly documented), but we should write code with assumption that we write good code, not a bad one.

[the remark this answers is not in the document]
Comment (max-vogler) [reply grr/server/grr_response_server/flows/general/osquery_test.py]: IMHO, both approaches make sense, both "public members first" and "c-style bottom-up definition". Since we do not have a guideline or style guide regarding this, I would favour accepting any sensible grouping of class members in this PR.

File: grr/server/grr_response_server/flows/general/osquery_test.py [hunk 11]
```diff
@@ -147,48 +152,108 @@ def testFailure(self):
...
+
+  def _RunQuery(self, query):
+    session_id = self._NewTestFlowStartAndRun(query)
     return flow_test_lib.GetFlowResults(self.client_id, session_id)
 
+  def _AssertTablesMatch(
+    self,
+    table: rdf_osquery.OsqueryTable,
+    expected: List[OrderedDict]
+  ) -> None:
```
[the remark this answers is not in the document]
Comment (simstoykov) [reply grr/server/grr_response_server/flows/general/osquery_test.py:188]: Ultimately I think it's a trade-off, and it's a matter of taste which one is better. 
Benefits of "AssertTablesMatch":
1. To me the test bodies are more readable with "AssertTablesMatch" as opposed to having a few assertions. 
2. If a test fails, the stack trace will include "AssertTablesMatch", so it is obvious that the problem is in the tables. Whereas with plain assertions we get no semantic context.

Downsides:
1. If a test fails, you need to look into the implementation of AssertTablesMatch, and figure out what is failing.
I'm biased but it's not difficult to decode `self.assertEqual(table_col.name, expected_col_name)` or `self.assertEqual(table_col_values, expected_col_vlues)`. (Jokes on me, I have a mistype)

I don't feel too strongly about it, though, so I can revert your code to how it used to be.

For mine, however, I will need a minimum of 10 additional assertions in the non-truncating test, and 11 in the other one. And what if we decide to increase the truncating length to 20? I need the automation.

What are your thoughts?

File: grr/server/grr_response_server/flows/general/osquery_test.py (hunk 6 above)
Comment (simstoykov) [reply grr/server/grr_response_server/flows/general/osquery_test.py:163]: Another thing is that the `[{ "foo": "quux", "baz": "thud", "bar": "norf" }] * 42` sample input wouldn't capture errors such as reordering of rows, as all of them are the same.

File: grr/server/grr_response_server/flows/general/osquery_test.py [hunk 12]
```diff
@@ -147,48 +152,108 @@ def testFailure(self):
...
+  ]
 
+
+class FakeOsqueryFlowTest(flow_test_lib.FlowTestsBaseclass):
   def setUp(self):
     super(FakeOsqueryFlowTest, self).setUp()
     self.client_id = self.SetupClient(0)
 
-  def _RunQuery(self, query):
+  def _NewTestFlowStartAndRun(self, query):
```
[the remark this answers is not in the document]
Comment (simstoykov) [reply grr/server/grr_response_server/flows/general/osquery_test.py:171]: Sorry I overlooked - I didn't rename `_RunQuery`, I separated some of its functionality in `_NewTestFlowStartAndRun`. But it's true that it wasn't obvious to me in the beginning that it starts a new flow. Does renaming it to `_RunQueryInNewFlow` sound good to you, or is it too wordy?

In general the scope of the two functions is not big, so I agree that maybe shorter names would be a better fit. I just prefer to be more explicit in the names, so that the code reads like a story. What do you think would be the best names for the two functions?

File: grr/server/grr_response_server/flows/general/osquery.py (hunk 7 above)
Comment (panhania) [reply grr/server/grr_response_server/flows/general/osquery.py:41]: In our codebase the convention is to use `TODO` comments. The format is `TODO(username)` (where `username` is the person that created or should deal with the issue at some point) and `TODO(link)` where `link` is a hyperlink to a bug report or an issue.

File: grr/server/grr_response_server/flows/general/osquery.py (hunk 9 above)
Comment (panhania) [reply grr/server/grr_response_server/flows/general/osquery.py:72]: Yeah, in that case it should stay as it is. Resolved.

File: grr/server/grr_response_server/flows/general/osquery_test.py (hunk 6 above)
Comment (panhania) [reply grr/server/grr_response_server/flows/general/osquery_test.py:163]: > However, what if a test fails because some cell values disagree?

If values are relevant they should not be introduced next to the place where you make assertions, otherwise you make assumption on what the format the helper is using, forcing the reader to familiarize themselves with the helper code (which is bad for the reasons explained in the next paragraph).

> Isn't it readable enough if the complexity is hidden behind helper functions?

If the reader has to jump around the code in order to understand what it is happening, it is not readable. Tests should read like linear scenario and be obvious. Here, I had to spend significant amount of time to understand what this helper is doing.

Moreover, it is not only about the readability. The more complex the helper is, the more likely you are to end up in a scenario where your generated data is invalid, e.g. it is empty and your tests passes trivially. Ultimately, in such cases, you should write tests for the helper itself (and at this point you not only introduce complexity, but also you have written more code than just inlining some input in the test code itself).

> Another thing is that the [{ "foo": "quux", "baz": "thud", "bar": "norf" }] * 42 sample input wouldn't capture errors such as reordering of rows, as all of them are the same.

Tests should be minimal and written for a particular purpose. If you want to test whether your code preserves the order, you should write small test case with pre-made order-sensitve input. If you want to test whether the code works with large input, then you can do `[single_row] * 42` and it still tests what it should.

File: grr/server/grr_response_server/flows/general/osquery_test.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (panhania) [reply grr/server/grr_response_server/flows/general/osquery_test.py:188]: > To me the test bodies are more readable with "AssertTablesMatch" as opposed to having a few assertions.

I would not count it as a benefit, because it is not objective. For me it is less readable and forces me to execute code in my head and jump around the code instead of reading in in a linear way.

> If a test fails, the stack trace will include "AssertTablesMatch", so it is obvious that the problem is in the tables. Whereas with plain assertions we get no semantic context.

The name of the test case should be everything that you need. You are testing tables, so obviously the problem will be with the tables. If you test more then one thing, then your test is wrong and should be split.

> I don't feel too strongly about it, though, so I can revert your code to how it used to be.

I kind of to, so I would prefer to revert to the old code.

> For mine, however, I will need a minimum of 10 additional assertions in the non-truncating test, and 11 in the other one

10 or 11 is within reason, IMO.

>  And what if we decide to increase the truncating length to 20? I need the automation.

That's kind of the problem here that you depend on an implementation detail. Truncating length should be a parameter and in the tests you should cover two scenarios:

  * Use some small truncating length to write tests with particular values that are asserted.
  * Use some big truncating length to assert have some vague assertion that the "shape" is still preserved.
